software/spmd/bsg_attr_noalias_ex/saxpy-c.c

```diff
@@ -0,0 +1,132 @@
+#include <bsg_manycore.h>
+#include "saxpy-c.h"
+
+/*
+ * saxpy_c result: Without bsg_attr_noalias, the compiler cannot guarantee that
+ * stores to C do not affect A and B, so load/store instructions are not
+ * reordered to overlap loop iterations and hide load latency.
+ *
+ * The disassembly and compiler flags are shown in snippets/saxpy_c.md
+ */
+void saxpy_c(float  *  A, float  *  B, float *C, float alpha) {
+        bsg_unroll(4)
+        for(int i = 0;  i < N_ELS; ++i) {
+                C[i] = alpha * A[i] + B[i];
+        }
+}
+
+
+/*
+ * saxpy_c_noalias: With bsg_attr_noalias, the code "guarantees" that stores to
+ * C do not affect A and B so load/store instructions can be reordered because
+ * the pointers do not alias.
+ *
+ * The disassembly and compiler flags are shown in snippets/saxpy_c_noalias.md
+ *
+ */
+void saxpy_c_noalias(float * bsg_attr_noalias A, float * bsg_attr_noalias B, float * bsg_attr_noalias C, float alpha) {
+        bsg_unroll(4)
+        for(int i = 0;  i < N_ELS; ++i) {
+                C[i] = alpha * A[i] + B[i];
+        }
+}
+
+
+/*
+ * Use caution when applying bsg_attr_noalias to multi-dimensional
+ * pointers. In this example LLVM produces worse code than GCC.
+
+ * This is a known issue in LLVM with a patch under review (as of 8/20):
+ * Google: restrict support in llvm
+ * LLVM Patch: https://reviews.llvm.org/D69542
+
+ * In this example, A is a 2xN_ELS array. We will perform saxpy using
+ * A[0] as x and A[1] as y.
+
+ * In saxpy_noalias_A, below, bsg_attr_noalias is only applied to the first
+ * dimension of A. The compiler still believes that writes to C affect
+ * data in the second dimension of A and cannot reorder instructions.
+
+ * In saxpy_A_noalias, below, bsg_attr_noalias is only applied to the
+ * second dimension of A. similar to saxpy_noalias_A, the compiler
+ * sill believes that writes to C can affect the pointer in the first
+ * dimension of A and cannot reorder instructions
+
+ * bsg_attr_noalias must be applied to each pointer dimension to be
+ * inferred correctly.
+
+ * saxpy_noalias_A result: The compiler cannot guarantee that stores
+ * to C do not affect the second dimension of A, so loop iterations are
+ * not overlapped
+
+ * saxpy_A_noalias result: The compiler cannot guarantee that stores
+ * to C do not affect the first dimension of A, so loop iterations are
+ * not overlapped
+
+ * saxpy_noalias_noalias result: In LLVM, the compiler does not infer
+ * correct aliasing. GCC does (see below)
```

Comment:
This is probably because LLVM has basic support of restrict compared to GCC. GCC probably supports most of what "C" standard says. Along those lines, a good addition could be a test with restrict on struct members. If GCC supports most of what C standard says, GCC should better hide load latency while accessing pointers that are struct members.